<h1 style="text-align: center">
<div style="color: #DD3403; font-size: 60%">MicroProject #8</div>
<span style="">Building a Scene Recognition Model form Video Frames by Wade Fagen-Wlmschneider</span>


<hr style="color: #DD3403;">

## Data Source: Frames of a Video

Visual images are an important part of all media and Data Scientists are often using images as data sources.  In this MicroProject, you will create a simple model to detect the amount of time spent in two different "scenes". These "scenes" are taken from the office-hour style videos that you have been seeing in your different Modules. 

In doing this, you will be building a **simple artificial intelligence (AI) algorithm** that will be able to predict the classification of unknown future data.  

<hr style="color: #DD3403;">

## Part 1: Loading a Video Frame

For this MicroProject, we already did a screen capture from the DISCOVERY video [*"Outliers Impact on Correlation (m6-02b)"*](https://www.youtube.com/watch?v=bd6hQ2UcIJc) that is used as part of our [DISCOVERY lecture covering Correlation](https://discovery.cs.illinois.edu/learn/Towards-Machine-Learning/Correlation/).  We captured one video frame every second and stored it as an image, and those files are available fo you in the `frames` directory of this MicroProejct.

To analyze these images, the `skimage` library is commonly used to load image data into Python.  Specifically the `skimage.io.imread(filename)` will read a filename and return the pixel color for every pixel in the image.

To use the `imread` function, you will need to either do one of the following:

1. Either import all of `skimage` by using the import line `import skimage`.  After importing all of `skimage`, you will call the function using its fully qualified name: `skimage.io.imread(filename)`.

**OR**

2. Import only the `imread` function by using the import line `from skimage.io import imread`.  After importing only `imread`, you will call the function directly: `imread(filename)`

### Part 1.1: Read Pixel Data for `frames/frame_0001.jpg`

In the following cell, store the pixel color data from the file named `frames/frame_0001.jpg` in the variable `pixels` by using the `imread` function:


In [2]:
from skimage.io import imread

pixels = imread(r"C:\Users\WOW\Downloads\Microproject 8\Microproject 8\frames\frame_0001.jpg")

In [3]:
### TEST CASE for Part 1: Loading a Video Frame
tada = "\N{PARTY POPPER}"

assert("pixels" in vars()), "Make sure you store the pixel data in the variable `pixels`."
assert(pixels.shape == (360, 640, 3)), "Make sure you are getting the pixel color data from the correct file."
assert(pixels[0][0][0] == 91), "Make sure you are getting the pixel color data from the correct file."

print(f"{tada} All Tests Passed! {tada}")

🎉 All Tests Passed! 🎉


Here's the image you loaded into `pixels` for reference as we work on the next section:

![frame_0001](frames/frame_0001.jpg "frame_0001")


<hr style="color: #DD3403;">

## Part 2: Finding Average Image Colors

The **shape** of your data is the `rows` by `columns` by `color values` as 3-dimensional list.  Here's a formatted view of your `pixels` data:

```py
[
  [ [91, 83, 80], [91, 83, 80], [91, 83, 80] ], ... ],   # Row #1
  [ [91, 83, 80], [91, 83, 80], [91, 83, 80] ], ... ],   # Row #2
  ...                                                    # ...
]
```

The current shape of `pixels` is 360 rows by 640 columns by 3 colors.  Each of the three colors represent the three color channels on a screen: red, green, and blue.

Using `pixel.mean()`, we find the average color by grouping **ALL** the color channels (combining blues and reds and greens together).  Try it out:


In [4]:
pixels.mean()

np.float64(72.18011863425926)

We can also find the average of each color channel instead of grouping them all together.  The `pixels.reshape(-1, 3).mean(axis=0)` function will return the mean of each color channel across all pixels.  Check out the new mean output:

In [5]:
# pixels.reshape(-1, 3) is being used to arrange the `pixels` data into 1 column
# while still preserving the original color channel data for each pixel, which is
# necessary to get the mean for each color channel:
pixels.reshape(-1, 3)

array([[ 91,  83,  80],
       [ 91,  83,  80],
       [ 91,  83,  80],
       ...,
       [162, 131, 110],
       [162, 131, 110],
       [162, 131, 110]], shape=(230400, 3), dtype=uint8)

We can combine it with `.mean(axis=0)` to find the average color of the three color channels:

In [ ]:
pixels.reshape(-1, 3).mean(axis=0)

array([88.65917535, 67.45620226, 60.4249783 ])

Finally, when a `list` or `array` is returned, we can spread the values into different variables where each variable takes one value from the list.  For example:

> ```py
> # Spreads the [2, 4], so x=2, and y=4:
> x, y = [2, 4]
> ```

We can spread any size list into as many variables as elements in the list:

> ```py
> # Spreads a list of five elements into five variables:
> a, b, c, d, e = [20, 30, 40, 50, 60]
> # a = 20; b = 30; c = 40; d = 50; and e = 60.
> ```

This can be helpful for the average pixel colors of an image:

> ```py
> # Spreads the list of three color channels into r, g, and, b:
> r, g, b = pixels.reshape(-1, 3).mean(axis=0)
> ```

### Part 2.1: Finding the Average Color of One Image

Store `pixels`'s average red value in `r`, average green value in `g`, and average blue value in `b`:

In [7]:
r,g,b = pixels.reshape(-1,3).mean(axis=0)
print(r)

88.65917534722222


In [8]:
### TEST CASE for Part 2.1: Finding the Average Color of One Image
tada = "\N{PARTY POPPER}"

import math
assert("r" in vars()), "The average red value should be stored in the variable `r`."
assert("g" in vars()), "The average green value should be stored in the variable `g`."
assert("b" in vars()), "The average blue value should be stored in the variable `b`."
assert(math.isclose(r, 88.65917534722222, abs_tol=1)), f"Your average red value is incorrect (r={r})."
assert(math.isclose(g, 67.45620225694445, abs_tol=1)), f"Your average green value is incorrect (g={g})."
assert(math.isclose(b, 60.42497829861111, abs_tol=1)), f"Your average blue value is incorrect (b={b})."

print(f"{tada} All Tests Passed! {tada}")
print(f"- The image's average red color channel is {round(r)} / 255")
print(f"- The image's average green color channel is {round(g)} / 255")
print(f"- The image's average blue color channel is {round(b)} / 255")


🎉 All Tests Passed! 🎉
- The image's average red color channel is 89 / 255
- The image's average green color channel is 67 / 255
- The image's average blue color channel is 60 / 255


### Part 2.2: Finding the Average Color of All Images

The following code loops through every file in the `frames` directory -- this will include `frame_0001.jpg` (like you analyzed already) and also `frame_0002.jpg`, `frame_0003.jpg`, and all 300+ frames!

Create a DataFrame named `df` where each row is one frame with the following four columns:
- `frame`, the filename of the frame
- `r`, the average red color of the frame
- `g`, the average green color of the frame
- `b`, the average blue color of the frame

The structure of the code should be nearly identical to writing a simulation.  Instead of creating random variables for your real world data, your real world data will be the filename, and the average color values. 

- See: https://discovery.cs.illinois.edu/learn/Simulation-and-Distributions/Simple-Simulations-in-Python/

*(Hint: Start at the very beginning of this MicroProject and observe the steps taken to get the average colors of `frame_0001.jpg`.)*

In [23]:
import glob
import os
import pandas as pd

data = []
frames = []
for frameFileName in glob.glob(os.path.join("frames", "*.jpg")): 
  # `frameFileName` contains the filename of the frame (ex: "frames/frame_0001.jpg").
  # Use `frameFileName` for `imread` to read the frame image data.
  pixels = imread(frameFileName)
  data.append(pixels.reshape(-1, 3).mean(axis=0))
  frames.append(frameFileName)
  

df = pd.DataFrame(data, columns = ["r","g","b"])
df["frame"] = frames
df


r          g          b                  frame
0    88.659175  67.456202  60.424978  frames\frame_0001.jpg
1    88.697865  67.453529  60.475660  frames\frame_0002.jpg
2    88.028351  66.913845  60.064592  frames\frame_0003.jpg
3    88.825629  67.340347  60.491645  frames\frame_0004.jpg
4    88.211714  66.979661  59.983173  frames\frame_0005.jpg
..         ...        ...        ...                    ...
325   7.470391   7.473355   7.479188  frames\frame_0326.jpg
326   7.469779   7.472743   7.478576  frames\frame_0327.jpg
327   7.480234   7.481519   7.487826  frames\frame_0328.jpg
328   7.480004   7.481289   7.487595  frames\frame_0329.jpg
329   4.657478   4.658776   4.665082  frames\frame_0330.jpg

[330 rows x 4 columns]

In [24]:
### TEST CASE for Part 2.2: Finding the Average Color of All Images
tada = "\N{PARTY POPPER}"

import math
assert("df" in vars()), "Make sure your DataFrame is named `df`."
assert(len(df) == 330), "Your DataFrame has the incorrect number of rows."
assert("r" in df), "Your `df` is missing the `r` column."
assert("g" in df), "Your `df` is missing the `g` column."
assert("b" in df), "Your `df` is missing the `b` column."
assert("frame" in df), "Your `df` is missing the `frame` column."
assert( abs( df[ df.frame.str.endswith("_0001.jpg") ]["r"].sum() - 88 ) < 1 ), "You have calculated the color averages incorrectly."

print(f"{tada} All Tests Passed! {tada}")

🎉 All Tests Passed! 🎉


<hr style="color: #DD3403;">

## Part 3: Create a Simple Classifier

In the DISCOVERY lecture videos, there are two primary "scenes" in the video:

1. **"Office Hours Studio Scene"**, where Karle and Wade are talking to each other and the audience

2. **"Notebook Scene"**, where the notebook is displayed

View the `frames` folder on your computer and find **at least three more frames** that are in the "Office Hours Studio Scene" and **at least three more frames** that are in the "Notebook Scene".  Add the frames you found to the lists below:

In [25]:
# List of at least four office hour frames by the filename's frame number:
office_hour_frames = [1, 12, 26, 97 ]

# List of at least four notebook frames by the filename's frame number:
notebook_frames = [30, 45, 68, 85 ]

### Part 3.1: Observing the Average Colors of Your Frames

The following code uses your sample frames to display the average color values for your selected frames:

In [26]:
import os

print("== Office Hour Frames ==")
print( df[ df["frame"].isin( [os.path.join("frames", f"frame_{frame:04d}.jpg") for frame in office_hour_frames] ) ] )
print()
print("== Notebook Frames ==")
print( df[ df["frame"].isin( [os.path.join("frames", f"frame_{frame:04d}.jpg") for frame in notebook_frames] ) ] )

== Office Hour Frames ==
             r          g          b                  frame
0    88.659175  67.456202  60.424978  frames\frame_0001.jpg
11  106.565577  53.084271  45.153090  frames\frame_0012.jpg
25   88.111723  67.775538  59.760907  frames\frame_0026.jpg
96   88.675165  68.549384  62.152448  frames\frame_0097.jpg

== Notebook Frames ==
             r           g           b                  frame
29  237.225595  236.513451  236.777122  frames\frame_0030.jpg
44  233.249470  232.604392  232.842834  frames\frame_0045.jpg
67  231.294601  230.751567  231.001714  frames\frame_0068.jpg
84  230.724310  229.920130  230.498368  frames\frame_0085.jpg


### Part 3.2: Create Your Classifier Function

A **classifier function** is a function that takes data and gives a classification for that data.  Create a new function, `classifyFrame` that receives an `r`, `g`, and `b` value.

Using information from your frames above, have the function return the string `"office hour"` or `"notebook"` based on the values of `r`, `g`, and `b`. To do this, try to observe general trends in the color values for both kinds of frames.

**IMPORTANT**: Make sure your classifier can handle **ANY** input -- even frames you have not seen before!  For example, you might decide that you will call a frame an `"office hour"` frame if the sum of `r`, `g` and `b` is greater than 100 and otherwise it's a `"notebook"` scene.

In [31]:
def classifyFrame(r, g, b):
  # Return either "office hour" or "notebook" based on the values of `r`, `g`, and `b`.
  mean = (r+g+b)/3
  if mean >= 150:
    return "notebook"
  else: return "office hour"

Here's a function to test your classifier to make sure your function returns a valid result (feel free to edit this).
- Right now, focus on making sure your function returns a valid result (either `notebook` or `office hour`)
- In the next part, we'll work on making sure it is an accurate classifier.

In [32]:
# Testing the `classifyFrame` function:
classifyFrame(0, 0, 0)

'office hour'

In [33]:
### TEST CASE for Part 3.2: Create Your Classifier Function
tada = "\N{PARTY POPPER}"

r = classifyFrame(0, 0, 0)
assert(r == "notebook" or r == "office hour"), "Your classifier function is misclassifying frames."

r = classifyFrame(255, 255, 255)
assert(r == "notebook" or r == "office hour"), "Your classifier function is misclassifying frames."

r = classifyFrame(0, 255, 255)
assert(r == "notebook" or r == "office hour"), "Your classifier function is misclassifying frames."

r = classifyFrame(255, 255, 0)
assert(r == "notebook" or r == "office hour"), "Your classifier function is misclassifying frames."

print(f"{tada} All Tests Passed! {tada}")

🎉 All Tests Passed! 🎉


<hr style="color: #DD3403;">

## Part 4: Using Your Classifier!

Now that we have a classifier, we should run it on every frame and see if your classifier correctly classifies scenes in the video frames.

The following cell runs your `classifyFrame` classifier on every frame and adds a new column `scene`. Then, 20 random rows are displayed below.  This output can be checked against the actual images.
- You should see a mix of `office hour` and `notebook` values in your **scene** column.
- If your classifier is not correct, update your `classifyFrame` function in the previous section and re-run that cell, and then re-run this cell to use your updated `classifyFrame` function.

In [34]:
# The following cell runs your `classifyFrame` classifier on every frame and adds a new column `scene`:
df["scene"] = df.apply(lambda row: classifyFrame(row.r, row.g, row.b), axis=1)
df.sample(20)

r           g           b                  frame        scene
316  233.539870   66.995486   47.766740  frames\frame_0317.jpg  office hour
303   90.062201   70.626554   63.144909  frames\frame_0304.jpg  office hour
270  243.563937  242.652452  241.196849  frames\frame_0271.jpg     notebook
292  242.170039  241.025152  239.725022  frames\frame_0293.jpg     notebook
221  244.537031  243.870859  242.090972  frames\frame_0222.jpg     notebook
177  232.595569  231.870260  229.682127  frames\frame_0178.jpg     notebook
121   88.056259   68.120924   60.819306  frames\frame_0122.jpg  office hour
301   89.382860   70.110286   62.665534  frames\frame_0302.jpg  office hour
294   90.011541   70.339766   63.319740  frames\frame_0295.jpg  office hour
131   87.254870   67.507188   60.424058  frames\frame_0132.jpg  office hour
169   90.121641   70.037899   62.148294  frames\frame_0170.jpg  office hour
152  230.118242  229.249644  226.898051  frames\frame_0153.jpg     notebook
80   230.721385  229.915091  230.483030  frames\frame_0081.jpg     notebook
1     88.697865   67.453529   60.475660  frames\frame_0002.jpg  office hour
288  241.932990  240.723173  239.471111  frames\frame_0289.jpg     notebook
197  238.683503  237.989249  235.640690  frames\frame_0198.jpg     notebook
279  243.225564  242.176402  240.907209  frames\frame_0280.jpg     notebook
304   89.183216   69.996150   62.738594  frames\frame_0305.jpg  office hour
42   233.114714  232.519796  232.737435  frames\frame_0043.jpg     notebook
198  238.580994  237.763281  235.496502  frames\frame_0199.jpg     notebook

### Observing Your Results

In the next four cells, we display a frame and you'll run code to check what your classifier classified the frame as being!  Make sure to run the code for each frame.
- If your classier is off, you will need to update your `classifyFrame` function and re-run both the `classifyFrame` cella and the cell at the top of this section to generate a enw **scene** classification for each image.
- It may take several attempts to get this correct, building an AI can be difficult!

### Frame #0001: Office Hours

In [35]:
df[ df.frame.str.endswith("0001.jpg") ]

r          g          b                  frame        scene
0  88.659175  67.456202  60.424978  frames\frame_0001.jpg  office hour

![Frame 0001](frames/frame_0001.jpg)

### Frame #0081: Notebook

In [36]:
df[ df.frame.str.endswith("0081.jpg") ]

r           g          b                  frame     scene
80  230.721385  229.915091  230.48303  frames\frame_0081.jpg  notebook

![Frame 0001](frames/frame_0081.jpg)

### Frame #0191: Notebook

In [37]:
df[ df.frame.str.endswith("0191.jpg") ]

r           g           b                  frame     scene
190  233.117088  232.354644  230.103359  frames\frame_0191.jpg  notebook

![Frame 0001](frames/frame_0191.jpg)

### Frame #0306: Office Hours

In [38]:
df[ df.frame.str.endswith("0306.jpg") ]

r          g         b                  frame        scene
305  89.403867  70.149223  62.83901  frames\frame_0306.jpg  office hour

![Frame 0001](frames/frame_0306.jpg)

### Run the "Part 4" Test Case:


In [39]:
### TEST CASE for Part 4: Using Your Classifier
tada = "\N{PARTY POPPER}"
assert("scene" in df)
assert(len(df[ df.scene == "notebook" ]) > 100), "Your classifier function misclassified lots of frames. Take another look at how you are classifying `office hour` and `notebook` frames."
assert(len(df[ df.scene == "office hour" ]) > 75), "Your classifier function misclassified lots of frames. Take another look at how you are classifying `office hour` and `notebook` frames."
assert(len(df[ df.scene == "notebook" ]) + len(df[ df.scene == "office hour" ]) == len(df)), "Your classifier function misclassified lots of frames. Take another look at how you are classifying `office hour` and `notebook` frames."
assert( len( df[ (df.frame.str.endswith("0001.jpg")) & (df.scene == "office hour") ] ) == 1 ), "Your classifier function misclassified an 'office hours' frame. Take another look at how you are classifying `office hour` and `notebook` frames."
assert( len( df[ (df.frame.str.endswith("0306.jpg")) & (df.scene == "office hour") ] ) == 1 ), "Your classifier function misclassified an 'office hours' frame. Take another look at how you are classifying `office hour` and `notebook` frames."
assert( len( df[ (df.frame.str.endswith("0081.jpg")) & (df.scene == "notebook") ] ) == 1 ), "Your classifier function misclassified a 'notebook' frame. Take another look at how you are classifying `office hour` and `notebook` frames."
assert( len( df[ (df.frame.str.endswith("0191.jpg")) & (df.scene == "notebook") ] ) == 1 ), "Your classifier function misclassified a 'notebook' frame. Take another look at how you are classifying `office hour` and `notebook` frames."
print(f"{tada} All Tests Passed! {tada}")

🎉 All Tests Passed! 🎉


<hr style="color: #DD3403;">

## Part 5: Improve Your Classifier

There are some frames that are neither the `office hour` nor the `notebook`.  How did your classifier do?

Run the following cells to see how you classified some unexpected frames:

### Frame #0320: Data Science Duo Logo???

What did you classify the DUO logo as?  It's neither one, but we don't have that option!

In [40]:
df[ df.frame.str.endswith("0320.jpg") ]

r          g          b                  frame        scene
319  221.227565  71.838433  54.457305  frames\frame_0320.jpg  office hour

![Frame 0001](frames/frame_0320.jpg)

### Frame #328: Video Credits

What did you classify the video credits as?  It's another tricky one!


In [41]:
df[ df.frame.str.endswith("0328.jpg") ]

r         g         b                  frame        scene
327  7.480234  7.481519  7.487826  frames\frame_0328.jpg  office hour

![Frame 0328](frames/frame_0328.jpg)

### Part 5.1: Building a Second Classifier

Create a second classifier -- `classifyFrame2` -- that returns either `"notebook"`, `"office hour"` or `"other"`.  Your classifier should correctly handle the "Data Science Duo" (ex: #0320) frames and the "Credit" frames (ex: #0328).
- Think of what might **uniquely identify these two special frames** that is different than the office hour or notebook scenes.  The average color values displayed with the images above may help you with some ideas.
- You must still correctly classify `"notebook"` and `"office hour"` frames, in addition to classifying `"other"` frames
- *Hint: It may be helpful to first identify if it's an `"other"` since these scenes may be most distinctive.*

In [56]:
def classifyFrame2(r, g, b):
  # Return either "office hour", "notebook", or "other" based on the values of `r`, `g`, and `b`.
  mean = (r+g+b)/3
  if mean < 10: #credits
    return "other"
  elif (mean < 130) & (mean > 80): #DUO logo
    return "other"
  elif (mean >= 125 ):
    return "notebook"
  else: return "office hour"


### Part 5.2: Apply your `classifyFrame2` function

Using `classifyFrame2`, this code replaces the value in the column `scene` with your `classifyFrame2` classification function.  The output of this cell shows the last frames of the video, which we expect many of the results to be `"other"`:

In [ ]:
df["scene"] = df.apply(lambda row: classifyFrame2(row.r, row.g, row.b), axis=1)
df.tail(20)

r          g          b                  frame        scene
0    88.659175  67.456202  60.424978  frames\frame_0001.jpg  office hour
1    88.697865  67.453529  60.475660  frames\frame_0002.jpg  office hour
2    88.028351  66.913845  60.064592  frames\frame_0003.jpg  office hour
3    88.825629  67.340347  60.491645  frames\frame_0004.jpg  office hour
4    88.211714  66.979661  59.983173  frames\frame_0005.jpg  office hour
5   181.087739  52.186936  37.745208  frames\frame_0006.jpg        other
6   180.672951  52.104644  37.659015  frames\frame_0007.jpg        other
7   180.616185  52.125681  37.592253  frames\frame_0008.jpg        other
8    88.963446  66.552773  59.458294  frames\frame_0009.jpg  office hour
9   108.020807  53.174227  44.903989  frames\frame_0010.jpg  office hour
10  106.510043  53.178854  44.992795  frames\frame_0011.jpg  office hour
11  106.565577  53.084271  45.153090  frames\frame_0012.jpg  office hour
12  106.535226  52.983685  45.031502  frames\frame_0013.jpg  office hour
13  107.287148  53.664076  45.780673  frames\frame_0014.jpg  office hour
14  106.867882  53.494240  45.379852  frames\frame_0015.jpg  office hour
15  106.216827  52.878997  44.989566  frames\frame_0016.jpg  office hour
16  108.083845  53.682270  45.474601  frames\frame_0017.jpg  office hour
17   90.887626  65.819722  58.584336  frames\frame_0018.jpg  office hour
18   88.007270  67.002426  59.679549  frames\frame_0019.jpg  office hour
19   87.650078  66.533685  59.362396  frames\frame_0020.jpg  office hour

### Test Cases

You can re-run the sample image cells earlier in this notebook after running the code in Part 4.2 since we updated the DataFrame using your new classifier.  Remember to update `classifyFrame2` to update your new classifier (not the original one).

In [58]:
### TEST CASE for Part 5: Update Your Classifier to Account with an Other Category
tada = "\N{PARTY POPPER}"
assert("scene" in df)
assert(len(df[ df.scene == "notebook" ]) > 100), "There should be many frames classified as 'notebook'. There are frames being misclassified."
assert(len(df[ df.scene == "office hour" ]) > 75), "You should have more frames classified as 'office hour'. There are frames being misclassified."
assert(len(df[ df.scene == "other" ]) >= 15), "Most of the last 20 frames should be classified as 'other'. Try adjusting the bounds of your classification parameters."
assert(len(df[ df.scene == "other" ]) <= 18)   # Okay to classify the intro screens as well, but not any others.
assert(len(df[ df.scene == "notebook" ]) + len(df[ df.scene == "office hour" ]) + len(df[ df.scene == "other" ]) == len(df))
assert( len( df[ (df.frame.str.endswith("0001.jpg")) & (df.scene == "office hour") ] ) == 1 ), "Your classifier function misclassified an 'office hours' frame. Take another look at how you are classifying the four frame categories."
assert( len( df[ (df.frame.str.endswith("0306.jpg")) & (df.scene == "office hour") ] ) == 1 ), "Your classifier function misclassified an 'office hours' frame. Take another look at how you are classifying the four frame categories."
assert( len( df[ (df.frame.str.endswith("0081.jpg")) & (df.scene == "notebook") ] ) == 1 ), "Your classifier function misclassified a 'notebook' frame. Take another look at how you are classifying the four frame categories."
assert( len( df[ (df.frame.str.endswith("0191.jpg")) & (df.scene == "notebook") ] ) == 1 ), "Your classifier function misclassified a 'notebook' frame. Take another look at how you are classifying the four frame categories."
assert( len( df[ (df.frame.str.endswith("0317.jpg")) & (df.scene == "other") ] ) == 1 ), "Your classifier function misclassified an 'other' frame. Take another look at how you are classifying the four frame categories."
assert( len( df[ (df.frame.str.endswith("0325.jpg")) & (df.scene == "other") ] ) == 1 ), "Your classifier function misclassified an 'other' frame. Take another look at how you are classifying the four frame categories."
assert( len( df[ (df.frame.str.endswith("0328.jpg")) & (df.scene == "other") ] ) == 1 ), "Your classifier function misclassified an 'other' frame. Take another look at how you are classifying the four frame categories."
print(f"{tada} All Tests Passed! {tada}")

🎉 All Tests Passed! 🎉


<hr style="color: #DD3403;">

<hr style="color: #DD3403;">

## Publish you work to your GitHub

1.  ⚠️ **Make sure to save your work.** ⚠️ To do this, go to **File => Save All** or press **Crtl + S** on your keyboard. 
2.  After you have saved, exit this notebook and locate this file in your Module 6 folder. Move this file to your cloned GitHub repository folder which should be inside of the folder you created with your First and Last name. 
3. Open your terminal and change your directory to your cloned GitHub Repository and type in these commands:

4. git add -A
5. git commit -m "Any message here"
6. git push